### PR TITLE
added local backend and noop scenario

### PR DESCRIPTION
**Pull request by papagalu**

### CODE REVIEW

File: argus/backends/local/local_backend.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
+from argus.backends import base as base_backend
+from argus.backends import windows
+from argus import config as argus_config
+from argus import log as argus_log
+
+
+class LocalBackendMixin(windows.WindowsBackendMixin,
```
Comment (ader1990) [inline argus/backends/local/local_backend.py]: Name it LocalBackend, as it is no longer a mixin

File: argus/backends/local/local_backend.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+                        base_backend.BaseBackend):
+    """ Local Backend for testing windows """
```
Comment (ader1990) [inline argus/backends/local/local_backend.py]: .. for testing Windows instances that are already started and ready.

File: argus/backends/local/local_backend.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+    def __init__(self, name=None, userdata=None, metadata=None,
+                 availability_zone=None):
+        self._name = name
+        self._availability_zone = availability_zone
+
```
Comment (ader1990) [inline argus/backends/local/local_backend.py]: extra line

File: argus/backends/local/local_backend.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    def setup_instance(self):
+        pass
+
+    def cleanup(self):
+        pass
+
+    @property
+    def floating_ip(self):
+        return lambda: "10.7.1.206"
```
Comment (ader1990) [inline argus/backends/local/local_backend.py]: remove the hardcoded value here and use a config value (you have to create one if not there)

File: argus/backends/windows.py [hunk 5]
```diff
@@ -26,7 +26,7 @@ class WindowsBackendMixin(object):
 
     # pylint: disable=unused-argument
     def get_remote_client(self, username=None, password=None,
-                          protocol='http', **kwargs):
+                          protocol='https', **kwargs):
```
Comment (ader1990) [inline argus/backends/windows.py]: reimplement the method in the backend, so that you do not break other backends

File: argus/recipes/cloud/windows.py [hunk 6]
```diff
@@ -831,3 +831,21 @@ def prepare_cbinit_config(self, service_type):
...
+class CloudbaseInitNoopRecipe(CloudbaseinitRecipe):
+    def create_mock_metadata(self, service_type):
+        """Create the mocked metadata."""
+        pass
+
+    def delete_mock_metadata(self):
+        """Delete the mocked metadata."""
+        pass
+
+    def prepare_cbinit_config(self, service_type):
```
Comment (ader1990) [inline argus/recipes/cloud/windows.py]: take a look at other recipes and you can inherit one of them that implements the preparation as you desire

File: ci/tests.py [hunk 7]
```diff
@@ -276,3 +282,11 @@ class ScenarioPasswordLength(BaseWindowsScenario):
...
     test_classes = (test_smoke.TestPasswordLength,)
     recipe_type = recipe.CloudbaseinitPasswordRecipe
+
+
+class ScenarioNoError(BaseWindowsScenario):
+
+    backend_type = local_backend.LocalBackendMixin
+    test_classes = (smoke.TestNoError,)
+    recipe_type = recipe.CloudbaseInitNoopRecipe
+    service_type = util.CONFIG_DRIVE_SERVICE
```
Comment (ader1990) [inline ci/tests.py]: add a NO_SERVICE service type, where metadata service will have a noop implementation

File: argus/recipes/cloud/windows.py (hunk 6 above)
Comment (ader1990) [reply argus/recipes/cloud/windows.py]: or just use these two methods:
https://github.com/cloudbase/cloudbase-init-ci/blob/master/argus/recipes/cloud/windows.py#L301

**Review by ader1990 — CHANGES_REQUESTED** [review]

File: argus/recipes/cloud/windows.py [hunk 8]
```diff
@@ -831,3 +831,21 @@ def prepare_cbinit_config(self, service_type):
         self._cbinit_conf.set_conf_value(
             name="user_password_length",
             value="3")
+
+
+class CloudbaseInitNoopRecipe(CloudbaseinitRecipe):
+    def create_mock_metadata(self, service_type):
```
Comment (paulaCrismaru) [inline argus/recipes/cloud/windows.py]: Add new line above.

File: ci/tests.py [hunk 9]
```diff
@@ -30,13 +31,18 @@
...
 
 def _availability_zones():
-    api_manager = manager.APIManager()
     try:
+        api_manager = manager.APIManager()
         zones = api_manager.availability_zone_client.list_availability_zones()
         info = zones['availabilityZoneInfo']
         return {zone['zoneName'] for zone in info}
+    except:
+        return []
```
Comment (paulaCrismaru) [inline ci/tests.py]: Line 38 returns a dictionary. Here should be an empty dictionary.

**Review by paulaCrismaru — CHANGES_REQUESTED** [review]

File: argus/backends/base.py [hunk 10]
```diff
@@ -52,6 +52,8 @@ def __init__(self, name=None, userdata=None, metadata=None,
 
         self.userdata = userdata
         self.metadata = metadata
+        self._username = CONFIG.openstack.username
```
Comment (ader1990) [inline argus/backends/base.py]: this should be None, otherwise you add an openstack dependency to openstack

File: argus/backends/base.py [hunk 11]
```diff
@@ -52,6 +52,8 @@ def __init__(self, name=None, userdata=None, metadata=None,
...
+        self._password = CONFIG.openstack.password
```
Comment (ader1990) [inline argus/backends/base.py]: ditto (means same thing as before said)

File: argus/backends/local/local_backend.py [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+from argus import log as argus_log
+from argus import util
+
+CONFIG = argus_config.CONFIG
+
+
+class LocalBackend(windows_backend.WindowsBackendMixin,
+                   base_backend.BaseBackend,
+                   windows_backend.BaseMetadataProviderMixin):
+    """ Local Backend for testing windows machines
```
Comment (ader1990) [inline argus/backends/local/local_backend.py]: Local backend for ... Windows*

File: argus/backends/local/local_backend.py [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+        self.userdata = userdata
+        self.metadata = metadata
+        self._username = CONFIG.local.username
+        self._password = CONFIG.local.password
+
+    def get_remote_client(self, username=None, password=None,
+                          protocol='https', **kwargs):
+
+        return windows.WinRemoteClient(self.floating_ip(),
+                                       self._username, self._password,
```
Comment (ader1990) [inline argus/backends/local/local_backend.py]: send params by name

File: argus/config/openstack.py [hunk 14]
```diff
@@ -27,7 +27,7 @@ def __init__(self, config):
         super(OpenStackOptions, self).__init__(config, group="openstack")
         self._options = [
             cfg.StrOpt(
-                "image_ref", default=None, required=True,
```
Comment (ader1990) [inline argus/config/openstack.py:30]: leave it as it was, why did you need to change?

**Review by ader1990 — CHANGES_REQUESTED** [review]

File: argus/config_generator/windows/cb_init.py [hunk 15]
```diff
@@ -93,7 +94,11 @@ def _get_service(self, service_type):
             This can be HTTP, ConfigDrive, EC2, OpenNebula,
             CloudStack or MAAS.
         """
-        return '.'.join([util.SERVICES_PREFIX, self.SERVICES[service_type]])
+        if self.SERVICES[service_type]:
```
Comment (ader1990) [inline argus/config_generator/windows/cb_init.py]: you can remove the return/else and leave only
 if self.SERVICES[service_type]:
return '.'.join([util.SERVICES_PREFIX,
 +                             self.SERVICES[service_type]])

**Review by ader1990 — CHANGES_REQUESTED** [review]

File: argus/backends/base.py [hunk 16]
```diff
@@ -52,6 +52,8 @@ def __init__(self, name=None, userdata=None, metadata=None,
 
         self.userdata = userdata
         self.metadata = metadata
+        self._username = None
```
Comment (mateimicu) [inline argus/backends/base.py]: This is required only for local backend, no need to have it in the base class.

File: argus/backends/local/local_backend.py [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+
+
+class LocalBackend(windows_backend.WindowsBackendMixin,
+                   base_backend.BaseBackend,
+                   windows_backend.BaseMetadataProviderMixin):
+    """ Local Backend for testing Windows machines
+        that are already installed and ready"""
+    def __init__(self, name=None, userdata=None, metadata=None,
+                 availability_zone=None):
+        self._name = name
```
Comment (mateimicu) [inline argus/backends/local/local_backend.py]: You should use `super`, something like
```python
def __init__(self, name=None, userdata=None, metadata=None,
                    availability_zone=None):
    super(LocalBackend, self).__init__(name=name, userdata=userdata, metadata=metadata,
                                       availability_zone=availability_zone)
    self._username = CONFIG.local.username
    self._password = CONFIG.local.password
````

File: argus/backends/local/local_backend.py [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+        self._availability_zone = availability_zone
+        self.userdata = userdata
+        self.metadata = metadata
+        self._username = CONFIG.local.username
+        self._password = CONFIG.local.password
+
+    def get_remote_client(self, username=None, password=None,
```
Comment (mateimicu) [inline argus/backends/local/local_backend.py]: I think you can implement the `WindowsBackendMixin` redefine the `get_remote_client` and user `super` to pass the desired parameters.

File: argus/config/openstack.py (hunk 14 above)
Comment (mateimicu) [reply argus/config/openstack.py:30]: Because it's required we need to pass some "fake" date in order to start the Argus run, even if we don't use the OpenStack backend some date needs to be present in order to parse all the config options. But this should not be changed, for now, we just need a better way overall to `disable` a backend.

File: argus/config_generator/windows/cb_init.py [hunk 19]
```diff
@@ -34,7 +34,8 @@ class BasePopulatedCBInitConfig(base.BaseWindowsConfig):
         util.EC2_SERVICE: "ec2service.EC2Service",
         util.OPEN_NEBULA_SERVICE: "opennebulaservice.OpenNebulaService",
         util.CLOUD_STACK_SERVICE: "cloudstack.CloudStack",
-        util.MAAS_SERVICE: "maasservice.MaaSHttpService"
+        util.MAAS_SERVICE: "maasservice.MaaSHttpService",
+        util.NO_SERVICE: ""
```
Comment (mateimicu) [inline argus/config_generator/windows/cb_init.py]: add a comma at the end of this line.

File: argus/config_generator/windows/cb_init.py [hunk 20]
```diff
@@ -93,7 +94,11 @@ def _get_service(self, service_type):
             This can be HTTP, ConfigDrive, EC2, OpenNebula,
             CloudStack or MAAS.
         """
-        return '.'.join([util.SERVICES_PREFIX, self.SERVICES[service_type]])
+        if self.SERVICES[service_type]:
+            return
```
Comment (mateimicu) [inline argus/config_generator/windows/cb_init.py]: Because you return `None` the join at line 121 might fail, return an empty string.

File: argus/config_generator/windows/cb_init.py [hunk 21]
```diff
@@ -93,7 +94,11 @@ def _get_service(self, service_type):
             This can be HTTP, ConfigDrive, EC2, OpenNebula,
```
Comment (mateimicu) [inline argus/config_generator/windows/cb_init.py:94]: Document the new behaviour of this function

File: ci/tests.py [hunk 22]
```diff
@@ -30,13 +31,18 @@
 
 
 def _availability_zones():
-    api_manager = manager.APIManager()
     try:
+        api_manager = manager.APIManager()
         zones = api_manager.availability_zone_client.list_availability_zones()
         info = zones['availabilityZoneInfo']
         return {zone['zoneName'] for zone in info}
+    except:
```
Comment (mateimicu) [inline ci/tests.py]: Catch the specific exceptions that you want to catch, and maybe log a message on why did you suppress the exception.

File: ci/tests.py [hunk 23]
```diff
@@ -30,13 +31,18 @@
...
+        return {}
     finally:
-        api_manager.cleanup_credentials()
+        try:
```
Comment (mateimicu) [inline ci/tests.py]: Instead of adding a new `try - catch` block you could make `api_manager = None` at the start of the function, and just add an `if` statement in the `finally` block.

File: ci/tests.py (hunk 22 above)
Comment (mateimicu) [reply ci/tests.py]: And because you may want to log something it would be better to move this function under the `argus` directory

**Review by mateimicu — CHANGES_REQUESTED** [review]

File: argus/config/local.py [hunk 24]
```diff
@@ -0,0 +1,53 @@
...
+
+class LocalOptions(conf_base.Options):
+
+    """Config options available for the local setup."""
+
+    def __init__(self, config):
+        super(LocalOptions, self).__init__(config, group="local")
+        self._options = [
+            cfg.StrOpt(
+                "ip", default="", required=True,
```
Comment (stefan-caraiman) [inline argus/config/local.py]: Think it should be "id" (at least that's what the help comment indicates), also why have an empty default at all and required option at the same time? Seems redundant.

File: argus/config/local.py [hunk 25]
```diff
@@ -0,0 +1,53 @@
...
+                help="The id of the flavor that is to be used."),
+            cfg.StrOpt(
+                "username", default="CiAdmin", required=True,
```
Comment (stefan-caraiman) [inline argus/config/local.py]: Ditto for the "default" being present and also having a "required=True", seems redundant since the "required" will never trigger, maybe that could be refactored for all configs(small suggestion)

File: argus/config/local.py [hunk 26]
```diff
@@ -0,0 +1,53 @@
...
+                help="The default username existing on Argus-CI images used "
+                     "for connecting and interacting with the instance."),
+            cfg.StrOpt(
+                "password", default="Passw0rd", required=True,
```
Comment (stefan-caraiman) [inline argus/config/local.py]: Just struck me that we are not following best practices right here, having a possible login password as a default value.
Seems better to simply add it in the config file, just a future suggestion.

File: argus/backends/local/local_backend.py [hunk 27]
```diff
@@ -0,0 +1,46 @@
...
+from argus.client import windows
+from argus import config as argus_config
+from argus import log as argus_log
+from argus import util
+
+CONFIG = argus_config.CONFIG
+
+
+class LocalBackend(windows_backend.WindowsBackendMixin,
+                   base_backend.BaseBackend,
```
Comment (stefan-caraiman) [inline argus/backends/local/local_backend.py]: Slight(OCD) suggestion, wouldn't this look more readable?
`class LocalBackend(windows_backend.WindowsBackendMixin,
                                windows_backend.BaseMetadataProviderMixin,
                                base_backend.BaseBackend):`

**Review by stefan-caraiman — CHANGES_REQUESTED** [review]

File: argus/backends/local/local_backend.py (hunk 27 above)
Comment (mateimicu) [reply argus/backends/local/local_backend.py]: do you mean:
```python
class LocalBackend(windows_backend.WindowsBackendMixin,
                   windows_backend.BaseMetadataProviderMixin, base_backend.BaseBackend):
```
if yes then sure, it would look more readable, but will also be longer than 79 chars.

Comment (ader1990) [reply argus/backends/local/local_backend.py]: it s about the mixins, to be put first. I would go with that version, when the mixins are inherited first

Comment (stefan-caraiman) [reply argus/backends/local/local_backend.py]: Nope, just meant that it would be okay to have both classes from windows_backend under each other(like your formatting plus a third line for the base_backend), just seemed more readable to me that's all.

Comment (ader1990) [reply argus/backends/local/local_backend.py]: I'd go with what  @stefan-caraiman  @micumatei suggested

File: argus/backends/local/local_backend.py [hunk 28]
```diff
@@ -0,0 +1,46 @@
...
+                   windows_backend.BaseMetadataProviderMixin):
+    """ Local Backend for testing Windows machines
+        that are already installed and ready"""
```
Comment (ader1990) [inline argus/backends/local/local_backend.py]: that are running, have git installed and winrm configured.

File: argus/config/local.py [hunk 29]
```diff
@@ -0,0 +1,53 @@
+# Copyright 2016 Cloudbase Solutions Srl
```
Comment (ader1990) [inline argus/config/local.py]: 2017*

File: argus/config/local.py (hunk 24 above)
Comment (ader1990) [reply argus/config/local.py]: ip is okay

File: argus/config/local.py [hunk 30]
```diff
@@ -0,0 +1,53 @@
...
+class LocalOptions(conf_base.Options):
+
+    """Config options available for the local setup."""
+
+    def __init__(self, config):
+        super(LocalOptions, self).__init__(config, group="local")
+        self._options = [
+            cfg.StrOpt(
+                "ip", default="", required=True,
+                help="The id of the flavor that is to be used."),
```
Comment (ader1990) [inline argus/config/local.py]: change the config description

File: argus/config/local.py (hunk 25 above)
Comment (ader1990) [reply argus/config/local.py]: you can undefine it in the config, is okay to have required and default

File: argus/config/local.py (hunk 24 above)
Comment (ader1990) [reply argus/config/local.py]: here you don t need a default value

File: argus/config/local.py (hunk 26 above)
Comment (ader1990) [reply argus/config/local.py]: remove the default value

File: argus/config/local.py [hunk 31]
```diff
@@ -0,0 +1,53 @@
...
+            cfg.StrOpt(
+                "username", default="CiAdmin", required=True,
+                help="The default username existing on Argus-CI images used "
+                     "for connecting and interacting with the instance."),
+            cfg.StrOpt(
+                "password", default="Passw0rd", required=True,
+                help="The password for the default username."),
+            cfg.BoolOpt(
+                "require_sysprep", default=True,
```
Comment (ader1990) [inline argus/config/local.py]: no need for this option

File: argus/util.py [hunk 32]
```diff
@@ -65,6 +66,7 @@
     OPEN_NEBULA_SERVICE: "",
     MAAS_SERVICE: "",
     CONFIG_DRIVE_SERVICE: "",
+    NO_SERVICE: "",
```
Comment (ader1990) [inline argus/util.py]: no need for this line, as there is no need for metadata service url
also, make sure you don t have in the config files a separate section for the metadata url, if the metadata service type does not have an url defined

File: ci/tests.py [hunk 33]
```diff
@@ -30,13 +31,17 @@
 
 
 def _availability_zones():
```
Comment (ader1990) [inline ci/tests.py]: this method should be used only when having to deal with openstack backend

Comment (ader1990) [reply ci/tests.py]: but for now, we can suppress the exception and return an empty array

**Review by ader1990 — CHANGES_REQUESTED** [review]

File: ci/tests.py (hunk 33 above)
Comment (ader1990) [inline ci/tests.py]: make  this method a static property of class tempest manager. use caching, so that the output can be reused during the same run.

**Review by ader1990 — CHANGES_REQUESTED** [review]

File: argus/backends/local/local_backend.py [hunk 34]
```diff
@@ -0,0 +1,58 @@
+# Copyright 2017 Cloudbase Solutions Srl
```
Comment (ader1990) [inline argus/backends/local/local_backend.py:1]: - adds the possibility to use an already existing Windows
      machine that is running, has git installed,
      http winrm enabled and network connectivity.
    - with the NO_SERVICE metadata service we test that pre-metadata
      plugins have been successfully executed and that Cloudbase-Init does not error out.

File: argus/config_generator/windows/cb_init.py [hunk 35]
```diff
@@ -93,7 +94,9 @@ def _get_service(self, service_type):
...
             CloudStack or MAAS.
         """
-        return '.'.join([util.SERVICES_PREFIX, self.SERVICES[service_type]])
+        if self.SERVICES[service_type]:
```
Comment (ader1990) [inline argus/config_generator/windows/cb_init.py]: if self.SERVICES.get would be better as it will not error out if the key does not exist

**Review by ader1990 — CHANGES_REQUESTED** [review]

File: argus/util.py [hunk 36]
```diff
@@ -67,7 +68,6 @@
 
     OPEN_NEBULA_SERVICE: "",
     MAAS_SERVICE: "",
-    CONFIG_DRIVE_SERVICE: "",
```
Comment (mateimicu) [inline argus/util.py:70]: Why delete this?

File: ci/tests.py [hunk 37]
```diff
@@ -29,18 +29,7 @@
 from argus import util
 
 
-def _availability_zones():
```
Comment (micumatei) [inline ci/tests.py]: This could be preserved
